### ISSUE TITLE
Incorrect documentation for func-style

### ISSUE THREAD

**Issue by revolter** [eslint/eslint#12900]
**Tell us about your environment**

* **ESLint Version:**
* **Node Version:**
* **npm Version:**

**What parser (default, Babel-ESLint, etc.) are you using?**

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js

```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js

```

```bash

```

**What did you expect to happen?**

The correct code example for the configuration `/*eslint func-style: ["error", "expression"]*/` shouldn't have an arrow function, as they're disallowed by default.

https://eslint.org/docs/rules/func-style#expression

**What actually happened? Please include the actual, raw output from ESLint.**

```
/*eslint func-style: ["error", "expression"]*/

var foo = function() {
    // ...
};

var foo = () => {};
```

**Are you willing to submit a pull request to fix this bug?**

Maybe.

**Comment by platinumazure** [comment]
I can't reproduce this: Both examples are lint-free when run in the [demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgZnVuYy1zdHlsZTogW1wiZXJyb3JcIiwgXCJleHByZXNzaW9uXCJdKi9cblxudmFyIGZvbyA9IGZ1bmN0aW9uKCkge1xuICAgIC8vIC4uLlxufTtcblxudmFyIGZvbyA9ICgpID0+IHt9OyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==).

It might help to fill in the issue template and share your ESLint version. Also, please let me know if something is wrong with my demo example. Thanks!

**Comment by mdjermanovic** [comment]
Hi @revolter, thanks for the issue!

I can confirm that the example is actually correct. The `"expression"` option allows all function expressions and all arrow functions, regardless of the context. It just disallows function declarations.

> "allowArrowFunctions": true (default false) allows the use of arrow functions (honoured only when using declaration)

 "honoured only when using declaration" applies to the option, and means that this option changes the behavior of this rule only if the main string option is `"declaration"`.

I agree that this sentence can be misinterpreted. 

A PR to improve the documentation is welcome!

**Comment by platinumazure** [comment]
Should we disallow the `allowArrowFunctions: true` option when the user is using "expression" mode? It's not a great experience to have an option that doesn't do anything in some cases.

**Comment by mdjermanovic** [comment]
@revolter just to confirm that we understand this bug report:

```js
/*eslint func-style: ["error", "expression"]*/

var foo = () => {};
```

You think that `var foo = () => {};` should be in Examples of **incorrect** code?

**Comment by mdjermanovic** [comment]
> Should we disallow the `allowArrowFunctions: true` option when the user is using "expression" mode? It's not a great experience to have an option that doesn't do anything in some cases.

Someone might also expect that `["error", "expression", { "allowArrowFunctions": false }]` disallows arrow functions. 

Would it be a breaking change to remove a misleading option that doesn't affect the behavior?

**Comment by mdjermanovic** [comment]
Documentation for the `"declaration"` option could be also improved.

The option disallows only function expressions/arrow functions used to initialize a variable. It also never disallows arrow functions that use `this`.

```js
/*eslint func-style: ["error", "declaration", { "allowArrowFunctions": false }]*/

var foo = function () {}; // error

var bar = () => {}; // error

arr.map(function() {}); // ok

arr.map(() => {}); // ok

var baz = () => this.something; // ok
```

**Comment by revolter** [comment]
> You think that `var foo = () => {};` should be in Examples of **incorrect** code for this rule with the default `"expression"` option?

Yes.

**Comment by mdjermanovic** [comment]
Thanks!

The example is correct and the rule is in sync with the example.

That's indeed the intended behavior. If the main string option is set to `"expression"`, arrow functions are allowed.

(even if `"allowArrowFunctions"` is set to `false`, because it applies only to `"declaration"`) 

I agree that the following can be easily misinterpreted: 

> "allowArrowFunctions": true (default false) allows the use of arrow functions (honoured only when using declaration)

PR to clarify this option in the documentation is welcome!

**Comment by anikethsaha** [comment]
> Should we disallow the allowArrowFunctions: true option when the user is using "expression" mode? It's not a great experience to have an option that doesn't do anything in some cases.

I think `allowArrowFunctions` does nothing as arrow function cant be possible for `declaration`.
